src/roadmap/v1/upcoming.py

```diff
@@ -1,12 +1,81 @@
+import typing as t
+
+from datetime import date
+from functools import cache
+
 from fastapi import APIRouter
+from pydantic import AfterValidator
+from pydantic import BaseModel
+from pydantic import TypeAdapter
 
-from roadmap.data import UPCOMING_DATA
+from roadmap.common import ensure_date
+from roadmap.config import Settings
+from roadmap.models import Meta
 
 
 router = APIRouter(prefix="/upcoming-changes", tags=["Upcoming Changes"])
 
+Date = t.Annotated[str | date | None, AfterValidator(ensure_date)]
+
+
+class Upcoming(BaseModel):
+    ticket: str
+    title: str
+    summary: str
+    type: str
+    package: str
+    architecture: str
+    release: str
+    release_date: Date
+
+
+class UpcomingResponseDetails(BaseModel):
+    detailFormat: int
+    summary: str
+    potentiallyAffectedSystems: int
+    trainingTicket: str
+    dateAdded: Date
+    lastModified: Date
+
+
+class UpcomingResponse(BaseModel):
+    name: str
+    type: str
+    release: str
+    date: Date
+    details: UpcomingResponseDetails
+
+
+class WrappedUpcomingResponse(BaseModel):
+    meta: Meta
+    data: list[UpcomingResponse]
+
+
+@cache
+def get_upcoming_data():
+    with open(Settings.create().upcoming_json_path, "r") as file:
```

Comment:
Use the already created Settings instance.
```suggestion
    with open(SETTINGS.upcoming_json_path, "r") as file:
```

Reply:
If I do this I won't be able to test it. The SETTINGS are made at import time, so the monkeypatch comes after. Suggestions?

I agree I don't like this usage.

Reply:
On solution would be to add an optional path parameter and use that if provided, using `SETTINGS.upcoming_json_path` by default. Call this function directly in testing instead of using a TestClient and provide a path to the settings file.

Probably the correct way to do this is to switch to using dependency injection for the settings so that it can be [overridden during testing](https://fastapi.tiangolo.com/advanced/testing-dependencies/?h=over#overriding-dependencies-during-testing).

Reply:
I did some more digging and experimentation. It's [not possible](https://github.com/fastapi/fastapi/discussions/10503#discussioncomment-7367434) to use FastAPIs dependency injection machinery outside of a route function. I figured that was the case.

We could switch injecting a settings instance and passing it around. That worked pretty well in my testing.